core/lib/server/pub.ts

```diff
@@ -589,3 +600,574 @@ export const deletePub = async (pubId: PubsId) =>
+									join.on((eb) =>
+										eb.and([
+											eb("pubs.id", "=", eb.ref("pub_values.pubId")),
+											// eb("pub_values.relatedPubId", "is not", null),
+										])
+									)
+								)
+								.innerJoin("pub_fields", "pub_fields.id", "pub_values.fieldId")
+								.leftJoin("PubsInStages", "pubs.id", "PubsInStages.pubId")
+								.$if(Boolean(fieldSlugs), (qb) =>
+									qb.where("pub_fields.slug", "in", fieldSlugs!)
+								)
+								.select([
+									"pubs.id as pubId",
+									"pub_fields.schemaName as schemaName",
+									"pub_fields.slug as slug",
+									"pubs.pubTypeId",
+									"pubs.communityId",
+									"pubs.createdAt",
+									"pubs.updatedAt",
+									"PubsInStages.stageId",
+									"pub_values.id as valueId",
+									"pub_values.fieldId",
+									"pub_values.value",
+									"pub_values.relatedPubId",
+									"pub_values.createdAt as valueCreatedAt",
+									"pub_values.updatedAt as valueUpdatedAt",
+									"pubs.parentId",
+									// increment the depth
+									sql<number>`pub_tree.depth + 1`.as("depth"),
+									// this is a standard way to detect cycles
+									// we keep track of the path we've taken so far,
+									// and set a flag if we see a pub that is already in the path
+									// https://www.postgresql.org/docs/current/queries-with.html#QUERIES-WITH-CYCLE
+									sql<boolean>`pubs.id = any(pub_tree."path")`.as("isCycle"),
+									sql<PubsId[]>`array_append(pub_tree."path", pubs.id)`.as(
+										"path"
+									),
+								])
+								.where("pub_tree.depth", "<", depth)
+								.where("pub_tree.isCycle", "=", false)
+								.$if(cycle === "exclude", (qb) =>
+									// this makes sure we don't include the first pub that is part of a cycle
+									qb.where(sql<boolean>`pubs.id = any(pub_tree.path)`, "=", false)
+								)
+						)
+					)
+			)
+			// this CTE finds the top level pubs and limits the result
+			// counter intuitively, this is CTE is referenced in the above `withRecursive` call, despite
+			// appearing after it. This is allowed in Postgres. See https://www.postgresql.org/docs/current/sql-select.html#SQL-WITH
+			// this is mostly because `kysely` does not allow you to put a normal CTE before a recursive CTE
+			.with("root_pubs_limited", (cte) =>
+				cte
+					.selectFrom("pubs")
+					.selectAll("pubs")
+					.$if(Boolean(props.pubId), (qb) => qb.where("pubs.id", "=", props.pubId!))
+					.$if(Boolean(props.stageId), (qb) =>
+						qb
+							.innerJoin("PubsInStages", "pubs.id", "PubsInStages.pubId")
+							.where("PubsInStages.stageId", "=", props.stageId!)
+					)
+					.$if(Boolean(props.communityId), (qb) =>
+						qb.where("pubs.communityId", "=", props.communityId!)
+					)
+					.$if(Boolean(props.pubTypeId), (qb) =>
+						qb.where("pubs.pubTypeId", "=", props.pubTypeId!)
+					)
+					.$if(Boolean(limit), (qb) => qb.limit(limit!))
+					.$if(Boolean(offset), (qb) => qb.offset(offset!))
+			)
```

Comment:
the reason we split up finding the root pubs here is because we want to be able to properly limit them. If we did it in the `withRecursive` part (which doesn't really work due to https://github.com/kysely-org/kysely/issues/646), where we `leftJoin` with `pub_values`, this wouldn't work, as `pubs.leftJoin('pub_values')` yields `#pubs * #pubValuesPerPub` rows. 

Ex.

If we have two pubs, one with `croccroc:title` and `croccroc:description`, and one with only `croccroc:title`, we would get the following rows (not actual output, but similar)

```ts
[{ 
	id: '1',
	slug: 'croccroc:title',
	value: 'Title of Pub 1'
},
{ 
	id: '1',
	slug: 'croccroc:description',
	value: 'A description of pub 1'
},
{ 
	id: '2',
	slug: 'croccroc:title',
	value: 'Title of Pub 2'
}]

```

If we were to put `limit(2)` near there, we would only get the rows

```ts
[{ 
	id: '1',
	slug: 'croccroc:title',
	value: 'Title of Pub 1'
},
{ 
	id: '1',
	slug: 'croccroc:description',
	value: 'A description of pub 1'
}]

```

which are only the values of a single pub. 

So we need to limit before we `leftJoin`, thus this setup.